### PR TITLE
WIP: Test gzip support in RESTEasy Reactive

### PR DESCRIPTION
**Pull request by saumya1singh**
CC @FroMage :)

### CODE REVIEW

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addAsResource(new StringAsset(APP_PROPS), "application.properties")
+                    .addClasses(BeanRegisteringRouteUsingObserves.class));
+
+    @Test
+    public void test() throws Exception {
```
Comment (NetoDevel) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:39]: method nomenclature is too generic.
If you can make it clear what is being tested

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.vertx.ext.web.Router;
+
+public class GZipTest {
+
+    private static final String APP_PROPS = "" +
+            "quarkus.http.enable-compression=true\n";
```
Comment (NetoDevel) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:20]: what do you think of a test with the other way?

`quarkus.http.enable-compression = false`

**Review by netodevel — CHANGES_REQUESTED** [review]
@SaumyaSingh1,
I suggested some small changes.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java (hunk 2 above)
Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:20]: If it is **disabled**, should we really perform any sort of test? I don't think that's needed when is not even enabled :)

Comment (NetoDevel) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:20]: but how to ensure that the property is working correctly?
and how do you make sure that no future code from other developers doesn't affect this and go undetected because you didn't have any tests?

Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:20]: @FroMage WDYT

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                .header("content-length", Matchers.equalTo(Integer.toString(longString.length())))
+                .body(Matchers.equalTo(longString));
+    }
+
+    @ApplicationScoped
+    static class BeanRegisteringRouteUsingObserves {
+
+        public void register(@Observes Router router) {
+
+            router.route("/compression").handler(rc -> {
```
Comment (FroMage) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:57]: This is testing reactive routes, but we want to be testing RESTEasy Reactive :)

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Test
+    public void test() throws Exception {
+
+        RestAssured.given().get("/compression").then().statusCode(200)
+                .header("content-encoding", "gzip")
+                .header("content-length", Matchers.not(Matchers.equalTo(Integer.toString(longString.length()))))
+                .body(Matchers.equalTo(longString));
+
+        RestAssured.given().get("/nocompression").then().statusCode(200)
```
Comment (FroMage) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:46]: For good measure, we should probably also test that we can send compressed data to the server.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java (hunk 1 above)
Comment (FroMage) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:39]:
```suggestion
    public void testServerCompression() throws Exception {
```

**Review by FroMage — CHANGES_REQUESTED** [review]

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java (hunk 2 above)
Comment (FroMage) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:20]: Well, we should test the default behaviour, that's true, but because we know what the default is ATM, we want to test the default behaviour without any configuration, to make sure it won't be compressed by default.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java (hunk 3 above)
Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:57]: From available articles (https://quarkus.io/guides/reactive-routes) I got the reactive route part, I tried searching for resources around testing RR but didn't get. Maybe we are supposed to do something like ?https://github.com/quarkusio/quarkus/blob/85d574ec82b3bc4735ed5933abe6cfaecad6b1d7/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/resource/basic/SubResourceRequestFilterTest.java#L61

Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:57]: oh wait, perhaps I should be testing this against some RR endpoint

Comment (FroMage) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:57]: Yes, indeed.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    //        }
+
+    @Path("/test")
+    public static class TestCompression {
+
+        //TODO - complete logic in method
+        @Path("/compression")
+        @GET
+        public void registerCompression(@Observes Router router) {
+            router.route("/test/compression").handler(routingContext -> {
```
Comment (SaumyaSingh1) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:83]: I am confused about this part @FroMage. Registering the endpoint is ok but by using the route thing I am again testing reactive routes instead of RESTEasy Reactive, no?

**Comment by SaumyaSingh1** [comment]
https://github.com/quarkusio/quarkus/issues/16425#issuecomment-832519045

In order to automate the process of "stoping the compression" we want to create an annotation. 

Situations/conditions when we want to stop/disable compression - 

1. If the data/files are already compressed (can be checked by magic number concept, .zip or .gz begins with specific magic num or by extension?)

2. If there are images (.png or .gif )

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java (hunk 4 above)
Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java]: So here we want to test if the data we are sending is compressed or not. Right?

Or do we want to compress the files first before sending them? (but there will be no purpose of enabling gzip then, so not this one)

**Comment by FroMage** [comment]
> Situations/conditions when we want to stop/disable compression -
> If the data/files are already compressed (can be checked by magic number concept, .zip or .gz begins with specific magic num. OR by extension type?)
> If there are images (.png or .gif ) then we want to stop compression

Well, first I'd like to see you test what the current behaviour is: will binary types get compressed or not?

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        RestAssured.given().get("/test/compression").then().statusCode(200)
+                .header("content-encoding", "gzip")
+                .header("content-length", Matchers.not(Matchers.equalTo(Integer.toString(longString.length()))))
+                .body(Matchers.equalTo(longString));
+
+        RestAssured.given().get("/test/nocompression").then().statusCode(200)
+                .header("content-encoding", "identity")
+                .header("content-length", Matchers.equalTo(Integer.toString(longString.length())))
+                .body(Matchers.equalTo(longString));
+    }
```
[the remark this answers is not in the document]
Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:56]: image
It is failing with this info. 
For endpoint `/test/nocompression`

[the remark this answers is not in the document]
Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:56]: So in begining I am initializing APP_PROPS as true
```
private static final String APP_PROPS = "" +
            "quarkus.http.enable-compression=true\n";

```

perhaps that's why content encoding is always `gzip` even for Binary Data i.e Images

[the remark this answers is not in the document]
Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:56]: When I set `quarkus.http.enable-compression` to false , the content-length=24000 bytes which is far more than 130 bytes (for text) or 165 bytes(for image)

```
private static final String APP_PROPS = "" +
            "quarkus.http.enable-compression=false\n";
```
image

[the remark this answers is not in the document]
Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:56]: So perhaps the conclusion is it is trying to compress images as well

**Comment by SaumyaSingh1** [comment]
> Well, first I'd like to see you test what the current behaviour is: will binary types get compressed or not?

Seems yes.

**For images :**
1. When `quarkus.http.enable-compression` is **false** ->

image

2. When `quarkus.http.enable-compression` is **true** ->

image

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (FroMage) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:56]: Yes, probably, could you add that test to your PR?

**Comment by SaumyaSingh1** [comment]
Interesting, so now I am supposed to create an annotation  `@DisableCompression`. 
This annotation should automatically add the header `Transfer-Encoding: identity` to stop compression.

And we will add this annotation at the endpoint `test/nocompression`.

**Comment by FroMage** [comment]
Yes.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    public void testServerCompression() throws Exception {
+
+        RestAssured.given().get("/test/compression").then().statusCode(200)
+                .header("content-encoding", "gzip")
+                .header("content-length", Matchers.not(Matchers.equalTo(Integer.toString(longString.length()))))
+                .body(Matchers.equalTo(longString));
+
+        RestAssured.given().get("/test/nocompression").then().statusCode(200)
+                .header("content-encoding", "identity")
+                .header("content-length", Matchers.equalTo((long) (createImage().getData().getDataBuffer().getSize() * 4L)))
```
Comment (SaumyaSingh1) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:59]:
```
(long) (createImage().getData().getDataBuffer().getSize() * 4L)
```

Finding the `content-length` for `BufferedImage` is something I am not much sure about. 
I took reference from here - 
https://stackoverflow.com/questions/632229/how-to-calculate-java-bufferedimage-filesize?noredirect=1&lq=1

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        graphic.setColor(Color.yellow);
+        graphic.drawString("Quarkus RESTEasy Reactive", 50, 120);
+        graphic.dispose();
+
+        // Save as PNG
+        File file = new File("src/test/resources/imageForTest.png");
+        ImageIO.write(bufferedImage, "png", file);
+        return bufferedImage;
+    }
+}
```
Comment (SaumyaSingh1) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java:105]: The image I am creating in `createImage()` looks like this - 

image

Perhaps "saving the image" is optional.

**Comment by SaumyaSingh1** [comment]
I have created the annotations, but I am yet to understand and implement the logic that how to make my custom annotation set the header "Transfer-Encoding` to `identity`. 

Am I going in the right direction? @FroMage

**Comment by FroMage** [comment]
I think your annotation should live next to `ServerExceptionMapper`, and then you have to make `EndpointIndexer.createResourceMethod` understand that annotation, and pass the info that compression is disabled in `ResourceMethod` which is used by `RuntimeResourceDeployment.buildResourceMethod` in order to create the set of `ServerRestHandler` that are to be called for the endpoint.

So if the endpoint has compression disabled, we need a new handler that adds the required header after the endpoint is invoked, probably before this line:

```java
        addHandlers(handlers, method, info, HandlerChainCustomizer.Phase.AFTER_METHOD_INVOKE);
```

BTW, I'll be on PTO the next three days so take your time, or ask others for help on Zulip if you need to, but it's possible they'll also be on PTO.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java (hunk 7 above)
Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java]: Humm I thought it's better to generate an image only when this test is run. 

You are right, it will make code simpler. Instead of creating an image through code it is better to use a static image 💯

Comment (SaumyaSingh1) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/GZipTest.java]: Done. 

Original size - 2914 bytes
Compressed size - 2741 bytes
image

**Comment by SaumyaSingh1** [comment]
Sure, it's public holiday here as well (tomorrow ig). 

I will use your explanation ^ and try to understand, implement it.

**Comment by SaumyaSingh1** [comment]
>and then you have to make EndpointIndexer.createResourceMethod understand that annotation, and pass the info that compression is disabled in ResourceMethod

In order to do this ^ I tried thinking of these 2 ways : 


**1.**  Should I directly introduce the new annotation inside `createResourceMethod()` in `EndPointIndexer.java`
https://github.com/quarkusio/quarkus/blob/c88b990f3163785e0eaaa59acd98db4c8657793c/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java#L377  and pass the info that compression is disabled somewhere in a new `ResourceMethod`
https://github.com/quarkusio/quarkus/blob/c88b990f3163785e0eaaa59acd98db4c8657793c/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java#L500

OR
**2.**   In `QuarkusServerEndpointIndexer` https://github.com/quarkusio/quarkus/blob/2cbd177416f92b9f45abb34b6f01acb8cb0f4eb5/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/QuarkusServerEndpointIndexer.java#L50 override `createResourceMethod()` and set the properties in a way to disable compression.

**Comment by SaumyaSingh1** [comment]
While working on the previous filter and other tasks I was easily able to debug my code written in `DevConsoleProcessor` or other files by hitting the DEv-UI endpoints such as http://localhost:8080/q/dev/io.quarkus.quarkus-resteasy-reactive/endpoints.

How should I go about it here e.g to check breakpoints in `EndpointIndexer`
https://github.com/quarkusio/quarkus/blob/c88b990f3163785e0eaaa59acd98db4c8657793c/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java

**Comment by FroMage** [comment]
I would say option 1, because this is not specific to Quarkus.

**Comment by FroMage** [comment]
Breaking on the indexer requires you to run the build steps, but you can do that by adding a `QuarkusUnitTest` in `quarkus-resteasy-reactive-deployment` like the others there, and running that test will invoke the build steps and therefore your breakpoint.

**Comment by gastaldi** [comment]
Hi @saumya1singh, any chance you'll resume this soon?

**Comment by saumya1singh** [comment]
hey @gastaldi, after 2.2.5 I am focusing on the Elektra release (plus have to provide training sessions to new interns in prod) , so can't give much time on this issue :/

**Comment by gastaldi** [comment]
@saumya1singh no worries, just curious 😉

**Comment by mkouba** [comment]
For the record - it seems that vertx will try to compress the static resources as well, see the `StaticHandler.skipCompressionForMediaTypes()` and `StaticHandler.skipCompressionForSuffixes()` methods. And I think that we should make this configurable too.

**Comment by mkouba** [comment]
Hm, to be honest I don't like the idea of opt-out via `@DisableCompression` (and config properties for static resources). I do understand that we're limited by how vertx-web implements compression but still.. it's not very flexible.

In theory, we could introduce two annotations (`@Compressed` and `@Uncompressed`) and a property to specify the default behavior for endpoints if `quarkus.http.enable-compression=true`, e.g. something like `quarkus.resteasy-reactive.endpoints-compressed` (`false` by default?). Furthermore,  we could introduce a simple [`@ServerResponseFilter`](https://quarkus.io/guides/resteasy-reactive#request-or-response-filters) that would add the `Transfer-Encoding: identity` header if the resource method is annotated with `@Uncompressed` OR `quarkus.resteasy-reactive.endpoints-compressed=false` and the method is not annotated with `@Compressed`. @saumya1singh @FroMage WDYT?

**Comment by gsmet** [comment]
@geoand @cescoffier could we try to come up with a plan so that @mkouba can put it into motion.

As for the requirements, I think:
- we should be able to compress endpoints without compressing static resources (and reciprocally)
- we should be able to enable/disable compression (with a global flag and then opt-in/opt-out)
- we will need to make sure gzip works for Reactive REST Client - I'm not sure if we handle GZip there at the moment

I don't think we need fancy algorithm to start with. Let's keep it simple and we will see later if we can expand.

**Comment by cescoffier** [comment]
Sorry, I don't have much availability this week.

First, the static handler can skip compression for a set of media types or extensions. I think we should rely on this. 

Basically, I would recommend enabling compression by default and letting decide compression vs. _raw_ using the annotation proposed by @mkouba. I know it's a big change in terms of behavior, but I think it's reasonable, and the user can go back to the previous behavior by disabling the compression globally. 

So, in other words, I agree with @mkouba's plan. We would need to document how to skip compression for static files based on the media types (I'm not sure we expose these settings, so we should).

Note that compression is handled at the netty level. Basically, we add the netty handler compressing the response.  That's why it's global.

**Comment by gsmet** [comment]
> honestly I don't consider this terribly important fot 2.8

I tend to disagree with that. We had numerous reports/questions about usage of the GZip option and it's especially important if you're paying your bandwidth bills in a cloud environment.

**Comment by cescoffier** [comment]
It's a question of trade off as you use more cpu and so cannot handle the same level of concurrency (which means you may need more instances). 

I also wanted to add the vertx 4.3 is going to extend how is handled compression. That may give us more flexibility, however that's for June (and I do not believe we can wait)

**Comment by mkouba** [comment]
Ok, to sum it up:
- we should change the default so that `quarkus.http.enable-compression=true`
- introduce `@Compressed` and `@Uncompressed`
- intoruce a config property to skip compression for static resources; and maybe we should skip ALL by default (the only problem is that `StaticHandler.skipCompressionForMediaTypes()` and `StaticHandler.skipCompressionForSuffixes()` methods accept an explicit set and don't support wildcards or anything like that)
- (?) introduce a config property to enable/disable compression for a specific technology, e.g. `resteasy-reactive` resource methods via `quarkus.resteasy-reactive.compressed-endpoints`, `false` by default

This would allow a user to:
- by default: just add `@Compressed` to a specific resource method and other resource methods would remain uncompressed
- enable compression for all resource methods via `quarkus.resteasy-reactive.compressed-endpoints=true` and then opt-out via `@Uncompressed` where needed
- enable compression for some static resources
- disable the compression altogether with `quarkus.http.enable-compression=false`

**Comment by geoand** [comment]
> I tend to disagree with that. We had numerous reports/questions about usage of the GZip option and it's especially important if you're paying your bandwidth bills in a cloud environment.

Not that I have seen

**Comment by geoand** [comment]
> This would allow a user to:
> 
> * by default: just add `@Compressed` to a specific resource method and other resource methods would remain uncompressed
> * enable compression for all resource methods via `quarkus.resteasy-reactive.compressed-endpoints=true` and then opt-out via `@Uncompressed` where needed
> * enable compression for some static resources
> * disable the compression altogether with `quarkus.http.enable-compression=false`

Seems reasonable

**Comment by mkouba** [comment]
> Seems reasonable in theory, but I don't really understand how compression can be used on per resource basis.

Well, I can imagine a resource with methods that return large JSON payloads and also with a method `foo()` that returns a zip file. In this case, you'd want to `quarkus.resteasy-reactive.compressed-endpoints=true` and `foo()` annotated with `@Uncompressed`?

**Comment by geoand** [comment]
My question was mainly from an implementation point of view. So basically what you are proposing is that we have two different ways of compressing output, one general one based on Vert.x and one custom one?

**Comment by mkouba** [comment]
> My question was mainly from an implementation point of view. So basically what you are proposing is that we have two different ways of compressing output, one general one based on Vert.x and one custom one?

Yes, but maybe I'm just overthinking on this issue ;-). 

The thing is that if we merge the `quarkus.http.enable-compression` and `quarkus.resteasy-reactive.compressed-endpoints`; and  interpret `quarkus.http.enable-compression=true` as _"compress all endpoints in all technologies"_ (resteasy-reactive, reactive-routes) then you won't be able to disable compression for endpoints by default an opt-in via `@Compressed` (i.e. there's no need for this annotation). Which might be acceptable. I'm not sure.

**Comment by FroMage** [comment]
It could be useful to look at how web servers do it: https://httpd.apache.org/docs/2.4/mod/mod_deflate.html and https://docs.nginx.com/nginx/admin-guide/web-server/compression/

Apparently, Apache doesn't enable anything by default, but gives two documented options:

- Compress html, text, js, css
- Compress everything but images

Nginx by default will compress only html, but documents the option to compress also xml

From that, I gather that we could go the same way and provide an option for compression by mime types:

```properties
quarkus.http.enable-compression=text/html,text/plain,text/xml,text/css,text/javascript,application/javascript
```

I guess that would accept MIME wildcards too. This applies to static resources as well as endpoints. This seems to be what makes the most sense and is the most flexible, no?
Or, we make it two options:

```properties
quarkus.http.enable-compression=true
# this is the default value, when compression is enabled
#quarkus.http.enable-compression.for=text/html,text/plain,text/xml,text/css,text/javascript,application/javascript
```

Now, sure, we can also support an annotation to make it per endpoint, but IMO that's a "nice to have" and not as important as this sort of config.

**Comment by mkouba** [comment]
> From that, I gather that we could go the same way and provide an option for compression by mime types:

Well, I like the idea but the problem is that `StaticHandler` currently does not support the opt-in approach; i.e. you can only say _"enable compression and disable for a set of media types"_.

> I guess that would accept MIME wildcards too. 

I don't think the current `StaticHandlerImpl` supports this; it seems to use `Object.equals()`.

**Comment by mkouba** [comment]
> > From that, I gather that we could go the same way and provide an option for compression by mime types:
> 
> Well, I like the idea but the problem is that `StaticHandler` currently does not support the opt-in approach; i.e. you can only say _"enable compression and disable for a set of media types"_.

@cescoffier had an idea that we could add `Content-Encoding: identity` to every response automatically (if `quarkus.http.enable-compression=true`) and let the next route in the chain decide whether to remove this header (and enable compression). I've experimented with this approach and it seems to work. I will add a link to my WIP branch once the GitHub UI is in sync with my quarkus repository...

**Comment by mkouba** [comment]
FYI https://github.com/mkouba/quarkus/tree/issue-16425

**Comment by geoand** [comment]
Closing as this has been addressed in https://github.com/quarkusio/quarkus/pull/24558
